### PR TITLE
Limit number of pods spawned in SchedulerPredicates validates resourc…

### PR DESCRIPTION
**Pull request by pskrzyns**
Fixes https://github.com/kubernetes/kubernetes/issues/29190,
With this patch test should spawn at most 10 pods on the smallest node.

### CODE REVIEW

**Comment by pskrzyns** [comment]
cc @kubernetes/sig-scheduling @kubernetes/sig-testing

**Comment by gmarek** [comment]
I don't think this PR helps in any way. We're mostly using 1 or 2 core machines, so using 500 mCPUs as a request of the Pod means that we have at most 4 of them. If you need to run this test on larger Nodes we can expose the size of those Pods as a flag to make it configurable, but I haven't heard about such need yet.

**Comment by pskrzyns** [comment]
@gmarek  
I'm running tests on bare metal so my nodes have 48 cores so as mentioned in https://github.com/kubernetes/kubernetes/issues/29190 I have almost 100 pods per node and this is a problem in my case, cause instead of scheduler i'm testing cluster performance.
cc @spiffxp

**Comment by gmarek** [comment]
OK, let's make it a flag then.

**Comment by spiffxp** [comment]
@gmarek I'd really prefer to find a way to make the test smarter, and avoid passing yet one more flag or env var to e2e.test

My understanding is this is a conformance test, not a perf test, so we shouldn't be aiming to hit control plane limits w/ number of pods.

Instead of fixed (or user-specified) pod size + dynamically computed number of pods, why not have dynamically computed pod size + fixed number of pods?

**Comment by gmarek** [comment]
OK, makes sense. Why don't we have both: max number of pods and min pod size?

**Comment by gmarek** [comment]
@pskrzyns - can you update your PR according to my comment (have both min size of the Pod and max numer of Pods)?

**Comment by pskrzyns** [comment]
@gmarek 
To make sure that I understood correctly, we want to have max 10 pods but not smaller than 500 micro cores ?

**Comment by gmarek** [comment]
Yup - this way we'll create less pods on small clusters and big clusters.

File: test/e2e/scheduler_predicates.go [hunk 1]
```diff
@@ -262,11 +262,18 @@ var _ = framework.KubeDescribe("SchedulerPredicates [Serial]", func() {
 	// It assumes that cluster add-on pods stay stable and cannot be run in parallel with any other test that touches Nodes or Pods.
 	// It is so because we need to have precise control on what's running in the cluster.
 	It("validates resource limits of pods that are allowed to run [Conformance]", func() {
+		MAX_NUMBER_OF_PODS := int64(10)
```
Comment (gmarek) [inline test/e2e/scheduler_predicates.go]: Define as a constant and use camel case names (start from a small letter, as we don't want it to be exported)

File: test/e2e/scheduler_predicates.go [hunk 2]
```diff
@@ -262,11 +262,18 @@ var _ = framework.KubeDescribe("SchedulerPredicates [Serial]", func() {
...
+		MIN_POD_CAPACITY := int64(500)
```
Comment (gmarek) [inline test/e2e/scheduler_predicates.go]: minPodCPURequest

File: test/e2e/scheduler_predicates.go [hunk 3]
```diff
@@ -281,7 +288,12 @@ var _ = framework.KubeDescribe("SchedulerPredicates [Serial]", func() {
 		}
 
 		var podsNeededForSaturation int
-		milliCpuPerPod := int64(500)
+
+		milliCpuPerPod := nodeMaxCapacity / MAX_NUMBER_OF_PODS
+		if milliCpuPerPod < MIN_POD_CAPACITY {
+			milliCpuPerPod = MIN_POD_CAPACITY
+		}
+		framework.Logf("Calculated pod capacity %vm", milliCpuPerPod)
```
Comment (gmarek) [inline test/e2e/scheduler_predicates.go]: Using pod capacity:...

**Comment by gmarek** [comment]
Few nits, lgtm after they're fixed.

**Comment by pskrzyns** [comment]
@gmarek PTAL

**Comment by gmarek** [comment]
@k8s-bot unit test this issue: #29992

**Comment by gmarek** [comment]
@k8s-bot unit test this issue: #27462

**Pull request merged**